[SPARK-45524][PYTHON][SQL] Initial support for Python data source read API

allisonwang-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->


### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR adds a new Python data source API and adds initial execution support for data source read. 

Here is an overview of the Python data source API. Please see the code (`python/pyspark/sql/datasource.py`) for more details.
```python
class DataSource(ABC):
    def __init__(self, options: Dict):
        self.options = options

    @property
    def name(self) -> str:
        ...

    def schema(self) -> Union[StructType, str]:
        ...

    def reader(self, schema: StructType) -> "DataSourceReader":
        ...

    # TODO: SPARK-45525
    def writer(self, schema: StructType, saveMode: str) -> "DataSourceWriter": 
        ...
```
DataSourceReader (this PR):
```python
class DataSourceReader(ABC):
    @abstractmethod
    def read(self, partition) -> Iterator:
        ...

    def partitions(self) -> Iterator[Any]:
        ...
```

Note this PR does not support reading a Python data source using `spark.read.format(...).load()`. This will be supported in a subsequent PR: [SPARK-45639](https://issues.apache.org/jira/browse/SPARK-45639).

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  5. If you fix a bug, you can clarify why it is a bug.
-->
To make Spark more accessible to Python users. For more information, please see this [SPIP](https://docs.google.com/document/d/1oYrCKEKHzznljYfJO4kx5K_Npcgt1Slyfph3NEk7JRU/edit?usp=sharing).

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes. Users can use the new Python data source API to create their own data source.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
New UTs.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

allisonwang-db committed:
data source read api

allisonwang-db commented:
cc @HyukjinKwon @ueshin @cloud-fan @dtenedor

dtenedor reviewed (commented):
This is a great start! Just reviewed the API for the first pass, will do another review of the implementation in the next round.

ueshin commented on python/pyspark/sql/tests/test_python_datasource.py:
```diff
@@ -0,0 +1,85 @@
+        df = self.spark.read.format(TestDataSource).load()
+        assertDataFrameEqual(df, [Row(id=None, value=0), Row(id=None, value=1)])
+
+    # TODO: support read with schema
```
Shall we have a JIRA ID for TODO?
```py
# TODO(SPARK-xxx): ...
```

ueshin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonDataSourcePartitionExec.scala:
```diff
@@ -0,0 +1,89 @@
+case class PythonDataSourcePartitionExec(
+    output: Seq[Attribute],
+    partitions: Seq[Array[Byte]])
+    extends LeafExecNode with InputRDDCodegen with PythonSQLMetrics {
```
nit: style; 2 white spaces for `extends`.

allisonwang-db commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,168 @@
+        """
+        Returns a DataSourceReader instance for reading data.
+
+        This method is required for readable data sources. It will be called once during
```
Good pointed. Added in the docstring of the datasource class.

allisonwang-db commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,168 @@
+        Returns a list of partitions for this data source.
+
+        This method is called once during the physical planning stage in the Spark planner
+        to generate a list of partitions. Note, partition values must be serializable.
```
I am not sure how many people are familiar with the serialization protocol here. I can change `serializable` to `picklable` to make it clearer.

allisonwang-db commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,168 @@
+        This method is called once during the physical planning stage in the Spark planner
+        to generate a list of partitions. Note, partition values must be serializable.
+
+        The planner then creates an RDD for each partition. Each partition value will be
```
Good point. Will update.

allisonwang-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonDataSource.scala:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    // Receive the pickled 'read' function.
+    val pickledFunction: Array[Byte] = {
+      val obj = new Array[Byte](length)
+      dataIn.readFully(obj)
+      obj
+    }
```
Nice utility function!

allisonwang-db commented on python/pyspark/sql/readwriter.py:
```diff
@@ -303,6 +329,50 @@ def load(
         self.options(**options)
+
+        # Load a Python data source
+        if isinstance(self._format, Callable):
```
Actually, the `self._format` here is not an instance of `DataSource`, it's the DataSource class (`Type[DataSource]`). Maybe I can just check if it's instance of `Type`.

ueshin replied:
Ah, then, how about using `issubclass(self._format, DataSource)`?
- https://docs.python.org/3/library/functions.html#issubclass

dtenedor replied:
Sounds good to me!

allisonwang-db committed:
address comments

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PythonDataSourceStrategy.scala:
```diff
@@ -0,0 +1,68 @@
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.python.UserDefinedPythonDataSourceReadRunner
+
+object PythonDataSourceStrategy extends Strategy with Logging {
```
shall we make it a logical rule, similar to `V2ScanRelationPushDown`

HyukjinKwon commented on python/pyspark/sql/readwriter.py:
```diff
@@ -69,20 +83,26 @@ class DataFrameReader(OptionUtils):
         return DataFrame(jdf, self._spark)
 
-    def format(self, source: str) -> "DataFrameReader":
+    def format(self, source: Union[str, Type[DataSource]]) -> "DataFrameReader":
```
can we support a short name?

allisonwang-db replied:
Yes we should also support using short names. But we need to support registering the data source first. This is a TODO in the PR description: `4. Register a Python data source and use its name to reference it in .format()`

HyukjinKwon replied:
I actually think also we should remove the support of directly passing `DataSource`. This implementation can't work with short names also if we want to make it working with other languages.

allisonwang-db replied:
Ah you meant we can't have a Python data source object when using it in Scala right? For example this won't work: `val df = spark.read.format(MyPythonDataSource).load()`. This is a valid point. 

@HyukjinKwon just brainstorming here: in order to use the short name, a data source must be registered somewhere. Then we first need to have an API to register a data source, something like `spark.dataSource.register(...)` similar to UDF.  

But can we make it more Pythonic? What if I just want to import a Python data source from a package and directly use it without registering? Having this `DataSource` object we can easily allow this use case:

```python
from src.my.datasources import MyDataSource

df = spark.read.format(MyDataSource).load()
``` 

This is similar to UDFs, where we don't need to register them to use them in PySpark.

HyukjinKwon replied:
Yes, we should have a registration mechanism like we do in Scala https://github.com/databricks/spark-xml/blob/master/src/main/resources/META-INF/services/org.apache.spark.sql.sources.DataSourceRegister. 

Can we investigate how other Python libraries do? One way is using INI file like https://packaging.python.org/en/latest/specifications/entry-points/

HyukjinKwon replied:
1. Create a INI file either under Spark conf directory, or current working directory (maybe optionally define the location via a static configuration).
2. Read it in `DataSource.lookupDataSource`, and loads the class. We will likely need a wrapper class that inherits Scala Datasource classes.
3. The wrapper Datasource invokes the Python interface to read/write, etc.

allisonwang-db replied:
Discussed offline. We should make it work with short names instead of changing the API here. I will revert the change in this file and work on this part in the next PR. This PR will focus on the execution and API of the data source.

HyukjinKwon commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,184 @@
+        Returns a list of integers:
+
+        >>> def partitions(self):
+        >>>     return [1, 2, 3]
```
```suggestion
        ...     return [1, 2, 3]
```

and should update below too

allisonwang-db replied:
Good catch!

HyukjinKwon commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,168 @@
+        The planner then creates an RDD for each partition. Each partition value will be
+        passed to `read(partition)` to read the data from this data source.
+
+        If this method is not implemented, or returns an empty list, Spark will create one
```
I believe there is a usecase. For example, in Scala, we have `LocalScan` that doesn't require partitions.

dtenedor replied:
Yes, it's possible, but you can always make a UDTF for that use case. That's very simple, you just have to make a class like this:

```
class MyDataSource:
  def eval(self, arg, other):
    # yield rows here
```

The user-defined data source can also explicitly define exactly one partition.
Both options are easy and worth preventing the performance harm from not partitioning a large data source. Having seen the latter before, it often makes the query fail completely.

allisonwang-db replied:
The default partition value should be `None` if the method is not implemented (indicating 1 partition).  I am opting to make `def partitions()` **optional** to implement from a usability perspective, making it easier for new users to get started.

dtenedor replied:
If you want to do that, please comment very carefully in the `DataSourceReader` class docstring itself that performance is likely to be limited for large data sets if the `partitions` method is not implemented or returns one partition (but it is OK to skip it if the result data set is small). We will want to make sure that every reader of this abstract base class thinks about this in order to prevent unintentionally making large data sources without any parallelism.

HyukjinKwon commented on python/pyspark/sql/readwriter.py:
```diff
@@ -303,6 +330,40 @@ def load(
+            data_source = data_source_cls(self._options)
+
+            # Get schema of the data source
+            schema = self._schema or data_source.schema()
```
This is a bit hacky. For example, if we run this in Spark Connect, the schema inference would happen within Python Spark Connect client.

HyukjinKwon replied:
Also, I don't think we can support SQL usecases if we do this here?

allisonwang-db replied:
Yea if a Python data source is registered and used in SQL, we need to have a separate Python process to instantiate the Python class and get the schema.

allisonwang-db replied:
Indeed it's a bit hacky right now. I will give this further thought when supporting `spark.read.schema(...)` in SPARK-45559. Currently, it does not work with the `schema(...)` clause.

dtenedor replied:
It will be necessary to inspect the values of the provided options for each call in order to determine which schema to generate. This is analogous to the Python UDTF `analyze` call we designed which runs on the driver.

allisonwang-db replied:
The `def schema()` method here can access `self.options` to generate schema based on the user inputs.

HyukjinKwon commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,184 @@
+
+        It can reference the `options` field to infer the data source's schema when
+        users do not explicitly specify it. This method is invoked once when calling
+        `spark.read.format(...).load()` to get the schema for a data source read operation.
```
Let's use two backquotes so it can be actual code block in the documentation:

```suggestion
        ``spark.read.format(...).load()`` to get the schema for a data source read operation.
```

allisonwang-db replied:
Good to know!

dtenedor commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,184 @@
+    """
+    A base class for data sources.
+
+    This class represents a custom data source that allows for reading from and
```
```suggestion
    This class represents a custom data source that allows for reading from and/or
```

dtenedor commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,184 @@
+    writing to it. The data source provides methods to create readers and writers
+    for reading and writing data, respectively. At least one of the methods `reader`
+    or `writer` must be implemented by any subclass to make the data source either
+    readable or writable.
```
```suggestion
    readable or writable (or both).
```

dtenedor commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,184 @@
+    readable or writable.
+
+    After implementing this interface, you can start to load your data source using
+    `spark.read.format(...).load()` and save data using `df.write.format(...).save()`.
```
```suggestion
    `spark.read.format(...).load()` and save data using `df.write.format(...).save()`.
    You can also create tables using SQL with:
    `CREATE TABLE tableName(<columnList>) USING <dataSourceName> OPTIONS <options>`
```

allisonwang-db replied:
This is nice but currently not doable :) I've created a new ticket here to keep track: [SPARK-45597](https://issues.apache.org/jira/browse/SPARK-45597)

dtenedor replied:
Cool, thanks for making a separate Jira for it :)

dtenedor commented on python/pyspark/sql/readwriter.py:
```diff
@@ -303,6 +330,40 @@ def load(
+            # Get schema of the data source
+            schema = self._schema or data_source.schema()
+            if isinstance(schema, str):
+                dt = _parse_datatype_string(schema)
```
we will want to check for parsing failures here and return a reasonable error message.

allisonwang-db replied:
Yup. Added a TODO to make schema robust: [SPARK-45559](https://issues.apache.org/jira/browse/SPARK-45559)

dtenedor commented on python/pyspark/sql/worker/plan_data_source_read.py:
```diff
@@ -0,0 +1,131 @@
+
+def main(infile: IO, outfile: IO) -> None:
+    """
+    Plan Python data source read.
```
can we expand this a bit more to mention when this gets called during the query processing sequence, and what it is responsible for receiving as input and returning as output?

allisonwang-db replied:
Sure!

dtenedor commented on python/pyspark/sql/worker/plan_data_source_read.py:
```diff
@@ -0,0 +1,131 @@
+        # Receive the data source instance.
+        data_source = read_command(pickleSer, infile)
+        if not isinstance(data_source, DataSource):
+            raise PySparkRuntimeError(
```
in general we should give error classes to each of these. But it seems this should never happen, should we just turn it into an assert instead? Same for other `PySparkRuntimeError` cases below.

allisonwang-db committed:
use logical rule + address comments

dtenedor reviewed (commented):
@allisonwang-db I did a review over the whole PR. Note, I extended your `.sql` file framework for UDTF calls in https://github.com/apache/spark/pull/43356 to support `PySparkRuntimeErorr`s and reasonable output in the `.sql.out` file (by stripping out filenames, line numbers, etc.) We could use a similar approach for data sources to make testing easier, I find the `.sql` file testing framework to be easy to use to make sure our test coverage is strong enough.

allisonwang-db commented on python/pyspark/sql/readwriter.py:
```diff
@@ -69,20 +82,26 @@ class DataFrameReader(OptionUtils):
     def __init__(self, spark: "SparkSession"):
         self._jreader = spark._jsparkSession.read()
         self._spark = spark
+        self._format: Optional[Union[str, Type[DataSource]]] = None
+        self._schema: Optional[Union[str, StructType]] = None
+        self._options: Dict[str, "OptionalPrimitiveType"] = dict()
```
@HyukjinKwon do you think we should define a separate class (like PyDataSourceReader) that contains all the logic related to Python data source? So that we don't need to expose these variables in the DataFrameReader class.

HyukjinKwon replied:
All these have to be removed. In theory, `DataFrameReader` should not have any change, and the Python sources are only referenced by `format` short name.

So yes I would like to separate them somewhere to make it easier to remove.

allisonwang-db replied:
Ok sounds good. I will refactor this in the next PR to make the review easier. Added a TODO.

cloud-fan commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,185 @@
+        Returns a DataSourceReader instance for reading data.
+
+        This method is required for readable data sources. It will be called once during
+        the physical planning stage in the Spark planner.
```
I think it's better to expose too many internal details. Just say it will be called only once when reading the data source.

cloud-fan commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,185 @@
+        >>> def schema(self):
+        ...   return StructType().add("a", "int").add("b", "string")
+        """
+        ...
```
is it the same as `raise NotImplementedError`?

allisonwang-db replied:
This is equivalent to returning None.
```python
def test():
  ...
v = test()
v is None 
> True
```

cloud-fan commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,185 @@
+        ----------
+        partition : object
+            The partition to read. It must be one of the partition values returned by
+            ``partitions()`` or None if ``partitions()`` method is not implemented.
```
need to update this if `partitions()` is required.

cloud-fan commented on python/pyspark/sql/readwriter.py:
```diff
@@ -151,8 +173,10 @@ def schema(self, schema: Union[StructType, str]) -> "DataFrameReader":
         if isinstance(schema, StructType):
             jschema = spark._jsparkSession.parseDataType(schema.json())
             self._jreader = self._jreader.schema(jschema)
+            self._schema = schema
```
why do we put it in both if and else instead of putting it after if-else only once?

allisonwang-db replied:
I got a Mypy error. I will refactor the way to store this info: [SPARK-45600](https://issues.apache.org/jira/browse/SPARK-45600)

cloud-fan commented on python/pyspark/sql/worker/plan_data_source_read.py:
```diff
@@ -0,0 +1,131 @@
+        handle_worker_exception(e, outfile)
+        sys.exit(-1)
+
+    send_accumulator_updates(outfile)
```
is it really needed? This python process runs on the driver side.

allisonwang-db replied:
This is from `analyze_udtf.py` cc @ueshin

ueshin replied:
It's needed if the accumulator is allowed in the data source definition. 
The file will be executed in a different Python process from the Driver Python process.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/pythonLogicalOperators.scala:
```diff
@@ -78,6 +80,42 @@ case class PythonMapInArrow(
+/**
+ * Represents a list of Python data source partitions.
+ */
+case class PythonDataSourcePartition(
```
```suggestion
case class PythonDataSourcePartitions(
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala:
```diff
@@ -101,7 +102,8 @@ class SparkOptimizer(
     V2Writes.ruleName :+
-    ReplaceCTERefWithRepartition.ruleName
+    ReplaceCTERefWithRepartition.ruleName :+
+    V2ScanRelationPushDown.ruleName
```
shall we put `PlanPythonDataSourceScan` here as well?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonDataSourcePartitionExec.scala:
```diff
@@ -0,0 +1,97 @@
+
+  override def inputRDD: RDD[InternalRow] = rdd
+
+  // Input is InternalRow, has to be turned into UnsafeRows.
```
input is unsafe row, isn't it? We invoked the unsafe projection in `lazy val unsafeRows`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonDataSourcePartitionExec.scala:
```diff
@@ -0,0 +1,97 @@
+  override protected val createUnsafeProjection: Boolean = true
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    val numOutputRows = longMetric("numOutputRows")
```
I think this can be a driver-side metrics, as it equals to the number of partitions, which is known at the driver side.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonDataSourcePartitionExec.scala:
```diff
@@ -0,0 +1,97 @@
+    }
+  }
+
+  override def executeCollect(): Array[InternalRow] = {
```
Do we really need to overwrite these methods? `PythonDataSourcePartitionExec` will never be the root node that can trigger the fast collect code path.

allisonwang-db replied:
That's right. This node should never be a root node. I will remove these methods.

allisonwang-db commented on python/pyspark/sql/datasource.py:
```diff
@@ -0,0 +1,184 @@
+        Returns a DDL string:
+
+        >>> def schema(self):
+        >>>    return "a INT, b STRING"
```
Yes. This PR does not support `.schema(...)`. Here is the follow up task: [SPARK-45559](https://issues.apache.org/jira/browse/SPARK-45559). I will add more test cases for schema.

allisonwang-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PythonDataSourceStrategy.scala:
```diff
@@ -0,0 +1,68 @@
+        elementSchema = schema,
+        children = partitionPlan.output,
+        evalType = PythonEvalType.SQL_TABLE_UDF,
+        udfDeterministic = false,
```
Maybe we can add an API in the data source or the reader / writer class?

dtenedor replied:
Sure, this sounds good.

allisonwang-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonDataSource.scala:
```diff
@@ -0,0 +1,86 @@
+    val length = dataIn.readInt()
+    if (length == SpecialLengths.PYTHON_EXCEPTION_THROWN) {
+      val msg = PythonWorkerUtils.readUTF(dataIn)
+      throw QueryCompilationErrors.tableValuedFunctionFailedToAnalyseInPythonError(msg)
```
Will add some tests (and also change the name of the error)

allisonwang-db committed:
address comments + add tests

HyukjinKwon commented:
Can we address https://github.com/apache/spark/pull/43360#discussion_r1365227126 here? We can't incrementally improve in the current way when we add other language support, etc.

allisonwang-db committed:
address comments

allisonwang-db commented:
@HyukjinKwon @cloud-fan @dtenedor thanks for the review! I've addressed the comments and removed the change in `readwriter` from this PR. It will be addressed in the next PR: [SPARK-45639](https://issues.apache.org/jira/browse/SPARK-45639)

allisonwang-db committed:
fix tests and update metrics

dtenedor reviewed (approved):
The API and testing look good now, thanks for iterating with all the reviewers to get this started.

ueshin commented:
Thanks! merging to master.
